---
title: "R Notebook_The comparison of FGFR4 or CD276 CARs infitrated in JR IM xenografts for 11 days"
output: html_notebook
---

This is an [R Markdown](http://rmarkdown.rstudio.com) Notebook. When you execute code within the notebook, the results appear beneath the code. 

Try executing this chunk by clicking the *Run* button within the chunk or by placing your cursor inside it and pressing *Ctrl+Shift+Enter*. 


#load library used in this project
```{r}
library(Seurat)
library(dplyr)
library(ggplot2)
library(cowplot)
library(patchwork)
library(dsb)
library(tidyr)
library(sctransform)
library(tibble)
library(RColorBrewer)

install.packages("hdf5r")
library(hdf5r)

```

#First, load your data using Read10x_h5 function(from 02_PrimaryAnalysisOutput)
```{r}
setwd("../FGFR4_CD276_CAR_TIL_CITEseq_20220818/02_PrimaryAnalysisOutput/04_RawMatricesH5")


FGFR4.28HTM.BBz_CAR1_1.data <- Read10X_h5(filename = "Custom_3A11_only/SCAF2991_3A11_CD28HTM_BBz_1_1_raw_feature_bc_matrix.h5", use.names = TRUE, unique.features = TRUE)

FGFR4.28HTM.BBz_CAR1_2.data <- Read10X_h5(filename = "Custom_3A11_only/SCAF2992_3A11_CD28HTM_BBz_1_2_raw_feature_bc_matrix.h5", use.names = TRUE, unique.features = TRUE)


FGFR4.28HTM.28z_CAR2.data <- Read10X_h5(filename = "Custom_3A11_only/SCAF2993_3A11_CD28HTM_CD28z_2_raw_feature_bc_matrix.h5", use.names = TRUE, unique.features = TRUE)


CD276.8HTM.BBz_CAR3.data <- Read10X_h5(filename = "Custom_MGB_only/SCAF2994_MGB7H3_LH_3_raw_feature_bc_matrix.h5", use.names = TRUE, unique.features = TRUE)

FGFR4.28HTM.BBz_CD276.8HTM.BBz_BiCisCAR4.data <- Read10X_h5(filename = "Custom_3A11_MGB/SCAF2995_3A11_CD28HTM_BBz_BiCisCAR_4_raw_feature_bc_matrix.h5", use.names = TRUE, unique.features = TRUE)

FGFR4.28HTM.28z_CD276.8HTM.BBz_BiCisCAR5.data <- Read10X_h5(filename = "Custom_3A11_MGB/SCAF2996_3A11_CD28HTM_CD28z_BiCisCAR_5_raw_feature_bc_matrix.h5", use.names = TRUE, unique.features = TRUE)

```

# simplify features name and check them after them
```{r}
head(rownames(x = FGFR4.28HTM.28z_CAR2.data[["Antibody Capture"]]))

rownames(x = FGFR4.28HTM.BBz_CAR1_1.data[["Antibody Capture"]]) <- gsub(pattern = "_*TotalSeqC", replacement = "",  x = rownames(x = FGFR4.28HTM.BBz_CAR1_1.data[["Antibody Capture"]]))

rownames(x = FGFR4.28HTM.BBz_CAR1_2.data[["Antibody Capture"]]) <- gsub(pattern = "_*TotalSeqC", replacement = "",  x = rownames(x = FGFR4.28HTM.BBz_CAR1_2.data[["Antibody Capture"]]))

rownames(x = FGFR4.28HTM.28z_CAR2.data[["Antibody Capture"]]) <- gsub(pattern = "_*TotalSeqC", replacement = "",  x = rownames(x = FGFR4.28HTM.28z_CAR2.data[["Antibody Capture"]]))

rownames(x = CD276.8HTM.BBz_CAR3.data[["Antibody Capture"]]) <- gsub(pattern = "_*TotalSeqC", replacement = "",  x = rownames(x = CD276.8HTM.BBz_CAR3.data[["Antibody Capture"]]))

rownames(x = FGFR4.28HTM.BBz_CD276.8HTM.BBz_BiCisCAR4.data[["Antibody Capture"]]) <- gsub(pattern = "_*TotalSeqC", replacement = "",  x = rownames(x = FGFR4.28HTM.BBz_CD276.8HTM.BBz_BiCisCAR4.data[["Antibody Capture"]]))

rownames(x = FGFR4.28HTM.28z_CD276.8HTM.BBz_BiCisCAR5.data[["Antibody Capture"]]) <- gsub(pattern = "_*TotalSeqC", replacement = "",  x = rownames(x = FGFR4.28HTM.CD28z_CD276.8HTM.BBz_BiCisCAR5.data[["Antibody Capture"]]))

head(rownames(x = FGFR4.28HTM.BBz_CAR1_1.data[["Antibody Capture"]]))

tail(rownames(x = FGFR4.28HTM.28z_CD276.8HTM.BBz_BiCisCAR5.data[["Antibody Capture"]]))

```

# separate 2 assay matrix, keep 19 CITE-seq features including hashing tags(HTO5,HTO6)
```{r}
FGFR4.28HTM.BBz_CAR1_1.rna <- FGFR4.28HTM.BBz_CAR1_1.data[["Gene Expression"]]
FGFR4.28HTM.BBz_CAR1_1.adt <- FGFR4.28HTM.BBz_CAR1_1.data[["Antibody Capture"]]

FGFR4.28HTM.BBz_CAR1_2.rna <- FGFR4.28HTM.BBz_CAR1_2.data[["Gene Expression"]]
FGFR4.28HTM.BBz_CAR1_2.adt <- FGFR4.28HTM.BBz_CAR1_2.data[["Antibody Capture"]]


FGFR4.28HTM.28z_CAR2.rna <- FGFR4.28HTM.28z_CAR2.data[["Gene Expression"]]
FGFR4.28HTM.28z_CAR2.adt <- FGFR4.28HTM.28z_CAR2.data[["Antibody Capture"]]

CD276.8HTM.BBz_CAR3.rna <- CD276.8HTM.BBz_CAR3.data[["Gene Expression"]]
CD276.8HTM.BBz_CAR3.adt <- CD276.8HTM.BBz_CAR3.data[["Antibody Capture"]]

FGFR4.28HTM.BBz_CD276.8HTM.BBz_BiCisCAR4.rna <- FGFR4.28HTM.BBz_CD276.8HTM.BBz_BiCisCAR4.data[["Gene Expression"]]
FGFR4.28HTM.BBz_CD276.8HTM.BBz_BiCisCAR4.adt <- FGFR4.28HTM.BBz_CD276.8HTM.BBz_BiCisCAR4.data[["Antibody Capture"]]

FGFR4.28HTM.28z_CD276.8HTM.BBz_BiCisCAR5.rna <- FGFR4.28HTM.28z_CD276.8HTM.BBz_BiCisCAR5.data[["Gene Expression"]]
FGFR4.28HTM.28z_CD276.8HTM.BBz_BiCisCAR5.adt <- FGFR4.28HTM.28z_CD276.8HTM.BBz_BiCisCAR5.data[["Antibody Capture"]]

```

# change column names with -1 pattern to conditions, make sure they are matched in different matrix
```{r}
colnames(FGFR4.28HTM.BBz_CAR1_1.rna) <- gsub(pattern = "-1", replacement = "", x = colnames(x = FGFR4.28HTM.BBz_CAR1_1.rna))
colnames(FGFR4.28HTM.BBz_CAR1_1.adt) <- gsub(pattern = "-1", replacement = "", x = colnames(x = FGFR4.28HTM.BBz_CAR1_1.adt))

head(colnames(FGFR4.28HTM.BBz_CAR1_1.rna))
head(colnames(FGFR4.28HTM.BBz_CAR1_1.adt))

colnames(FGFR4.28HTM.BBz_CAR1_2.rna) <- gsub(pattern = "-1", replacement = "", x = colnames(x = FGFR4.28HTM.BBz_CAR1_2.rna))
colnames(FGFR4.28HTM.BBz_CAR1_2.adt) <- gsub(pattern = "-1", replacement = "", x = colnames(x = FGFR4.28HTM.BBz_CAR1_2.adt))

head(colnames(FGFR4.28HTM.BBz_CAR1_2.rna))
tail(colnames(FGFR4.28HTM.BBz_CAR1_2.adt))

colnames(FGFR4.28HTM.28z_CAR2.rna) <- gsub(pattern = "-1", replacement = "", x = colnames(x = FGFR4.28HTM.28z_CAR2.rna))
colnames(FGFR4.28HTM.28z_CAR2.adt) <- gsub(pattern = "-1", replacement = "", x = colnames(x = FGFR4.28HTM.28z_CAR2.adt))

head(colnames(FGFR4.28HTM.28z_CAR2.rna))
tail(colnames(FGFR4.28HTM.28z_CAR2.adt))

colnames(CD276.8HTM.BBz_CAR3.rna) <- gsub(pattern = "-1", replacement = "", x = colnames(x = CD276.8HTM.BBz_CAR3.rna))
colnames(CD276.8HTM.BBz_CAR3.adt) <- gsub(pattern = "-1", replacement = "", x = colnames(x = CD276.8HTM.BBz_CAR3.adt))

head(colnames(CD276.8HTM.BBz_CAR3.rna))
tail(colnames(CD276.8HTM.BBz_CAR3.adt))


colnames(FGFR4.28HTM.BBz_CD276.8HTM.BBz_BiCisCAR4.rna) <- gsub(pattern = "-1", replacement = "", x = colnames(x = FGFR4.28HTM.BBz_CD276.8HTM.BBz_BiCisCAR4.rna))
colnames(FGFR4.28HTM.BBz_CD276.8HTM.BBz_BiCisCAR4.adt) <- gsub(pattern = "-1", replacement = "", x = colnames(x = FGFR4.28HTM.BBz_CD276.8HTM.BBz_BiCisCAR4.adt))

head(colnames(FGFR4.28HTM.BBz_CD276.8HTM.BBz_BiCisCAR4.rna))
tail(colnames(FGFR4.28HTM.BBz_CD276.8HTM.BBz_BiCisCAR4.adt))


colnames(FGFR4.28HTM.28z_CD276.8HTM.BBz_BiCisCAR5.rna) <- gsub(pattern = "-1", replacement = "", x = colnames(x = FGFR4.28HTM.28z_CD276.8HTM.BBz_BiCisCAR5.rna))
colnames(FGFR4.28HTM.28z_CD276.8HTM.BBz_BiCisCAR5.adt) <- gsub(pattern = "-1", replacement = "", x = colnames(x = FGFR4.28HTM.28z_CD276.8HTM.BBz_BiCisCAR5.adt))

head(colnames(FGFR4.28HTM.28z_CD276.8HTM.BBz_BiCisCAR5.rna))
tail(colnames(FGFR4.28HTM.28z_CD276.8HTM.BBz_BiCisCAR5.adt))

```

# split HTO features to hto matrix
```{r}

rownames(FGFR4.28HTM.BBz_CAR1_1.adt)
FGFR4.28HTM.BBz_CAR1_1.hto <- FGFR4.28HTM.BBz_CAR1_1.adt[18:19,]

rownames(FGFR4.28HTM.BBz_CAR1_2.adt)
FGFR4.28HTM.BBz_CAR1_2.hto <- FGFR4.28HTM.BBz_CAR1_2.adt[18:19,]

FGFR4.28HTM.28z_CAR2.hto <- FGFR4.28HTM.28z_CAR2.adt[18:19,]

CD276.8HTM.BBz_CAR3.hto <- CD276.8HTM.BBz_CAR3.adt[18:19,]

FGFR4.28HTM.BBz_CD276.8HTM.BBz_BiCisCAR4.hto <- FGFR4.28HTM.BBz_CD276.8HTM.BBz_BiCisCAR4.adt[18:19,]

FGFR4.28HTM.28z_CD276.8HTM.BBz_BiCisCAR5.hto <- FGFR4.28HTM.28z_CD276.8HTM.BBz_BiCisCAR5.adt[18:19,]

rownames(FGFR4.28HTM.28z_CD276.8HTM.BBz_BiCisCAR5.hto)

```

# for FGFR4.CD28HTM.BBz_CAR1_1 and FGFR4.CD28HTM.BBz_CAR1_2, due to hto5 & hto6 staining contamination, and we will apply DSB Normalization apply DSB normalization for adt data, so using DSB Normalization recommend QC method to define background/empty droplets separately step I, create a metadata dataframe of simple qc stats for each droplet

```{r}
FGFR4.28HTM.BBz_CAR1_1.adtOnly <- FGFR4.28HTM.BBz_CAR1_1.adt[1:17,]
FGFR4.28HTM.BBz_CAR1_2.adtOnly <- FGFR4.28HTM.BBz_CAR1_2.adt[1:17,]

CAR1_1_rna_size = log10(Matrix::colSums(FGFR4.28HTM.BBz_CAR1_1.rna))
CAR1_1_prot_size = log10(Matrix::colSums(FGFR4.28HTM.BBz_CAR1_1.adtOnly))
CAR1_1_ngene = Matrix::colSums(FGFR4.28HTM.BBz_CAR1_1.rna > 0)
CAR1_1_mtgene = grep(pattern = "^MT-", rownames(FGFR4.28HTM.BBz_CAR1_1.rna), value = TRUE)
CAR1_1_propmt = Matrix::colSums(FGFR4.28HTM.BBz_CAR1_1.rna[CAR1_mtgene, ]) / Matrix::colSums(FGFR4.28HTM.BBz_CAR1_1.rna)
CAR1_1_md = as.data.frame(cbind(CAR1_1_propmt, CAR1_1_rna_size, CAR1_1_ngene, CAR1_1_prot_size))
CAR1_1_md$CAR1_1_bc = rownames(CAR1_1_md)


CAR1_2_rna_size = log10(Matrix::colSums(FGFR4.28HTM.BBz_CAR1_2.rna))
CAR1_2_prot_size = log10(Matrix::colSums(FGFR4.28HTM.BBz_CAR1_2.adtOnly))
CAR1_2_ngene = Matrix::colSums(FGFR4.28HTM.BBz_CAR1_2.rna > 0)
CAR1_2_mtgene = grep(pattern = "^MT-", rownames(FGFR4.28HTM.BBz_CAR1_2.rna), value = TRUE)
CAR1_2_propmt = Matrix::colSums(FGFR4.28HTM.BBz_CAR1_2.rna[CAR2_mtgene, ]) / Matrix::colSums(FGFR4.28HTM.BBz_CAR1_2.rna)
CAR1_2_md = as.data.frame(cbind(CAR1_2_propmt, CAR1_2_rna_size, CAR1_2_ngene, CAR1_2_prot_size))
CAR1_2_md$CAR1_2_bc = rownames(CAR1_2_md)

```

# Step Ib: Quality control and defining cell-containing and background droplets
```{r}
CAR1_1_p1 = ggplot(CAR1_1_md[CAR1_1_md$CAR1_1_rna_size > 0, ], aes(x = CAR1_1_rna_size)) + geom_histogram(fill = "dodgerblue") + ggtitle("CAR1_1_RNA library size \n distribution")
CAR1_1_p2 = ggplot(CAR1_1_md[CAR1_1_md$CAR1_1_prot_size> 0, ], aes(x = CAR1_1_prot_size)) + geom_histogram(fill = "firebrick2") + ggtitle("CAR1_1_Protein library size \n distribution")
cowplot::plot_grid(CAR1_1_p1, CAR1_1_p2, nrow = 1)

CAR1_2_p1 = ggplot(CAR1_2_md[CAR1_2_md$CAR1_2_rna_size > 0, ], aes(x = CAR1_2_rna_size)) + geom_histogram(fill = "dodgerblue") + ggtitle("CAR1_2_RNA library size \n distribution")
CAR1_2_p2 = ggplot(CAR1_2_md[CAR1_2_md$CAR1_2_prot_size> 0, ], aes(x = CAR1_2_prot_size)) + geom_histogram(fill = "firebrick2") + ggtitle("CAR1_2_Protein library size \n distribution")
cowplot::plot_grid(CAR1_2_p1, CAR1_2_p2, nrow = 1)
       
```

# Nextly,define a vector of background / empty droplet barcodes based on protein library size and mRNA content
```{r}
CAR1_1_background_drops = CAR1_1_md[CAR1_1_md$CAR1_1_prot_size < 2.4, ]$CAR1_1_bc
CAR1_1_negative_mtx_rawprot = FGFR4.28HTM.BBz_CAR1_1.adtOnly[, CAR1_1_background_drops] %>% as.matrix()

CAR1_2_background_drops = CAR1_2_md[CAR1_2_md$CAR1_2_prot_size < 2.4, ]$CAR1_2_bc
CAR1_2_negative_mtx_rawprot = FGFR4.28HTM.BBz_CAR1_2.adtOnly[, CAR1_2_background_drops] %>% as.matrix()

```

# define a vector of cell-containing droplet barcodes based on protein library size and mRNA content
```{r}
CAR1_1_positive_cells = CAR1_1_md[CAR1_1_md$CAR1_1_prot_size > 2.4 & CAR1_1_md$CAR1_1_ngene > 400 & CAR1_1_propmt < 0.25, ]$CAR1_1_bc
CAR1_1_cells_mtx_rawprot = FGFR4.28HTM.BBz_CAR1_1.adtOnly[, CAR1_1_positive_cells] %>% as.matrix()

CAR1_2_positive_cells = CAR1_2_md[CAR1_2_md$CAR1_2_prot_size > 2.4 & CAR1_2_md$CAR1_2_ngene > 400 & CAR1_2_propmt < 0.25, ]$CAR1_2_bc
CAR1_2_cells_mtx_rawprot = FGFR4.28HTM.BBz_CAR1_2.adtOnly[, CAR1_2_positive_cells] %>% as.matrix()

```

# Check: are the number of cells in line with the expected recovery from the experiment?
```{r}
length(CAR1_1_positive_cells)

length(CAR1_2_positive_cells)
```

# removing isotype control background for ADT data and correcting for per-cell technical factor as a covariate
```{r}
isotypes = c("M_IgG1", "M_IgG2a", "M_IgG2b")
CAR1_1_dsb_norm_prot = DSBNormalizeProtein(cell_protein_matrix = CAR1_1_cells_mtx_rawprot, empty_drop_matrix = CAR1_1_negative_mtx_rawprot, denoise.counts = TRUE, use.isotype.control = TRUE, isotype.control.name.vec = isotypes)
FGFR4.28HTM.BBz_CAR1_1.adtOnly.norm <- as.sparse(CAR1_1_dsb_norm_prot)

CAR1_2_dsb_norm_prot = DSBNormalizeProtein(cell_protein_matrix = CAR1_2_cells_mtx_rawprot, empty_drop_matrix = CAR1_2_negative_mtx_rawprot, denoise.counts = TRUE, use.isotype.control = TRUE, isotype.control.name.vec = isotypes)
FGFR4.28HTM.BBz_CAR1_2.adtOnly.norm <- as.sparse(CAR1_2_dsb_norm_prot)

```

# ridge plot protein features and remove some outliers
```{r}
CAR1_1_pmtx = t(CAR1_1_dsb_norm_prot) %>% as.data.frame()
prots = colnames(CAR1_1_pmtx)
index1 = prots[1]; index2 = prots[length(prots)]
CAR1_1_pmtx1 = CAR1_1_pmtx %>% gather(CAR1_1_prot, CAR1_1_dsb_normalized_value, index1:index2)

CAR1_2_pmtx = t(CAR1_2_dsb_norm_prot) %>% as.data.frame()
CAR1_2_pmtx1 = CAR1_2_pmtx %>% gather(CAR1_2_prot, CAR1_2_dsb_normalized_value, index1:index2)

```

# For visualization
```{r}
CAR1_ridge_aes = list(
  ggridges::geom_density_ridges2(show.legend = F, inherit.aes = T, scale = 2.5) ,
  ggridges::theme_ridges(font_size = 10, font_family = "Helvetica",center_axis_labels = T,  grid = TRUE, line_size = 0.2), 
  geom_vline(xintercept = 0, linetype="dashed", color = "black", size=1) ,
  theme(strip.background =element_rect(fill="white")),
  theme(strip.text = element_text(colour = 'black', size = 10, face = "bold")) ,
  theme(axis.text.x = element_text(size = 14, face = "bold")) ,
  theme(axis.text.y = element_text(size = 10, face = "bold")) 
  )
ggplot(CAR1_1_pmtx1 %>% filter(CAR1_1_dsb_normalized_value > -5 & CAR1_1_dsb_normalized_value < 50), 
            aes(x = CAR1_1_dsb_normalized_value , y = CAR1_1_prot, alpha = 1)) + CAR1_ridge_aes + 
  ggtitle("CAR1_1_normalized with region 2 \n as background \n  outlier drops removed ")

ggplot(CAR1_2_pmtx1 %>% filter(CAR1_2_dsb_normalized_value > -5 & CAR1_2_dsb_normalized_value < 50), 
            aes(x = CAR1_2_dsb_normalized_value , y = CAR1_2_prot, alpha = 1)) + CAR1_ridge_aes + 
  ggtitle("CAR1_2_normalized with region 2 \n as background \n  outlier drops removed ")
```

# create FGFR4.28HTM.BBz_CAR1_1 and FGFR4.28HTM.BBz_CAR1_2 seurat object
```{r}
FGFR4.28HTM.BBz_CAR1_1 <- CreateSeuratObject(counts = FGFR4.28HTM.BBz_CAR1_1.rna[, CAR1_1_positive_cells], project = "FGFR4.28HTM.BBz_CAR1_1")
dim(FGFR4.28HTM.BBz_CAR1_1)

FGFR4.28HTM.BBz_CAR1_2 <- CreateSeuratObject(counts = FGFR4.28HTM.BBz_CAR1_2.rna[, CAR1_2_positive_cells], project = "FGFR4.28HTM.BBz_CAR1_2")
dim(FGFR4.28HTM.BBz_CAR1_2)
```

# Add “ADT” into seurat object
```{r}
FGFR4.28HTM.BBz_CAR1_1[["ADT"]] <- CreateAssayObject(FGFR4.28HTM.BBz_CAR1_1.adtOnly[, colnames(x = FGFR4.28HTM.BBz_CAR1_1)])

FGFR4.28HTM.BBz_CAR1_2[["ADT"]] <- CreateAssayObject(FGFR4.28HTM.BBz_CAR1_2.adtOnly[, colnames(x = FGFR4.28HTM.BBz_CAR1_2)])

```

# now add the DSB normalized data back to the object (the singlets defined above as “object”)
```{r}
FGFR4.28HTM.BBz_CAR1_1[["ADT"]] = SetAssayData(object = FGFR4.28HTM.BBz_CAR1_1[["ADT"]], slot = "data", new.data = FGFR4.28HTM.BBz_CAR1_1.adtOnly.norm[, colnames(x = FGFR4.28HTM.BBz_CAR1_1)])

FGFR4.28HTM.BBz_CAR1_2[["ADT"]] = SetAssayData(object = FGFR4.28HTM.BBz_CAR1_2[["ADT"]], slot = "data", new.data = FGFR4.28HTM.BBz_CAR1_2.adtOnly.norm[, colnames(x = FGFR4.28HTM.BBz_CAR1_2)])

head(FGFR4.28HTM.BBz_CAR1_1[["ADT"]]@data)

```

# QC and filter data
```{r}
FGFR4.28HTM.BBz_CAR1_1[["percent.mt"]] <- PercentageFeatureSet(FGFR4.28HTM.BBz_CAR1_1, pattern = "^MT-")
hist(FGFR4.28HTM.BBz_CAR1_1$percent.mt)
mean(FGFR4.28HTM.BBz_CAR1_1$percent.mt)
plot3 = VlnPlot(FGFR4.28HTM.BBz_CAR1_1, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)

FGFR4.28HTM.BBz_CAR1_2[["percent.mt"]] <- PercentageFeatureSet(FGFR4.28HTM.BBz_CAR1_2, pattern = "^MT-")
hist(FGFR4.28HTM.BBz_CAR1_2$percent.mt)
mean(FGFR4.28HTM.BBz_CAR1_2$percent.mt)
plot4 = VlnPlot(FGFR4.28HTM.BBz_CAR1_2, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)

hist(FGFR4.28HTM.BBz_CAR1_1$nFeature_RNA)
VlnPlot(FGFR4.28HTM.BBz_CAR1_1, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)

hist(FGFR4.28HTM.BBz_CAR1_2$nFeature_RNA)
VlnPlot(FGFR4.28HTM.BBz_CAR1_2, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)

```

# Merge FGFR4.28HTM.BBz_CAR1_1 and FGFR4.28HTM.BBz_CAR1_2 seurat objects and then do filter together
```{r}
FGFR4.28HTM.BBz_CAR1 <- merge(FGFR4.28HTM.BBz_CAR1_1, y = FGFR4.28HTM.BBz_CAR1_2, project = "FGFR4.28HTM.BBz_CAR1", merge.data = TRUE)

GetAssayData(FGFR4.28HTM.BBz_CAR1)[1:10, 1:15]

DefaultAssay(FGFR4.28HTM.BBz_CAR1) <- "ADT"

GetAssayData(FGFR4.28HTM.BBz_CAR1)[1:10, 1:15]

head(FGFR4.28HTM.BBz_CAR1@meta.data)
tail(FGFR4.28HTM.BBz_CAR1@meta.data)

FGFR4.28HTM.BBz_CAR1
head(colnames(FGFR4.28HTM.BBz_CAR1))
tail(colnames(FGFR4.28HTM.BBz_CAR1))

table(FGFR4.28HTM.BBz_CAR1$orig.ident)

```

# QC for combined FGFR4.28HTM.BBz_CAR1 data
```{r}

DefaultAssay(FGFR4.28HTM.BBz_CAR1) <- "RNA"
FGFR4.28HTM.BBz_CAR1[["percent.mt"]] <- PercentageFeatureSet(FGFR4.28HTM.BBz_CAR1, pattern = "^MT-")
hist(FGFR4.28HTM.BBz_CAR1$percent.mt)
mean(FGFR4.28HTM.BBz_CAR1$percent.mt)
VlnPlot(FGFR4.28HTM.BBz_CAR1, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)

hist(FGFR4.28HTM.BBz_CAR1$nFeature_RNA)
hist(FGFR4.28HTM.BBz_CAR1$nCount_RNA)

```

# filter data for combined FGFR4.28HTM.BBz_CAR1 data
```{r}
FGFR4.28HTM.BBz_CAR1.sub <- subset(FGFR4.28HTM.BBz_CAR1, subset = nFeature_RNA > 400 & nFeature_RNA < 8000 & percent.mt < 15)
VlnPlot(FGFR4.28HTM.BBz_CAR1.sub, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)

dim(FGFR4.28HTM.BBz_CAR1.sub)

table(FGFR4.28HTM.BBz_CAR1.sub$orig.ident)
```

# save RDS data
```{r}
saveRDS(FGFR4.28HTM.BBz_CAR1.sub, file = "../SeuratAnalysisData/RDSdata/FGFR4.28HTM.BBz_CAR1_combined.sub.rds")

saveRDS(FGFR4.28HTM.BBz_CAR1, file = "../SeuratAnalysisData/RDSdata/FGFR4.28HTM.BBz_CAR1_combined.rds")
```

# For other 4 groups CART cells, due to there is hash5 and hash6 to distinct CAR1 and CAR2, we use HTO demultiplexing separately to define background/empty droplets. CLR normalization, MultiseqDemux(); firstly creat seurat object and add "HTO"data normalization HTO data first
```{r}
FGFR4.28HTM.28z_CAR2 <- CreateSeuratObject(counts = FGFR4.28HTM.28z_CAR2.rna, project = "FGFR4.28HTM.28z_CAR2")
FGFR4.28HTM.28z_CAR2[["HTO"]] <- CreateAssayObject(counts = FGFR4.28HTM.28z_CAR2.hto)

CD276.8HTM.BBz_CAR3 <- CreateSeuratObject(counts = CD276.8HTM.BBz_CAR3.rna, project = "CD276.8HTM.BBz_CAR3")
CD276.8HTM.BBz_CAR3[["HTO"]] <- CreateAssayObject(counts = CD276.8HTM.BBz_CAR3.hto)

FGFR4.28HTM.BBz_CD276.8HTM.BBz_BiCisCAR4 <- CreateSeuratObject(counts = FGFR4.28HTM.BBz_CD276.8HTM.BBz_BiCisCAR4.rna, project = "FGFR4.28HTM.BBz_CD276.8HTM.BBz_BiCisCAR4")
FGFR4.28HTM.BBz_CD276.8HTM.BBz_BiCisCAR4[["HTO"]] <- CreateAssayObject(counts = FGFR4.28HTM.BBz_CD276.8HTM.BBz_BiCisCAR4.hto)

FGFR4.28HTM.28z_CD276.8HTM.BBz_BiCisCAR5 <- CreateSeuratObject(counts = FGFR4.28HTM.28z_CD276.8HTM.BBz_BiCisCAR5.rna, project = "FGFR4.28HTM.28z_CD276.8HTM.BBz_BiCisCAR5")

FGFR4.28HTM.28z_CD276.8HTM.BBz_BiCisCAR5[["HTO"]] <- CreateAssayObject(counts = FGFR4.28HTM.28z_CD276.8HTM.BBz_BiCisCAR5.hto)

```

# Normalization and demultiplexing by MULTIseqDemux for each object independently; Due to MULTIseqDemux works when HTODemux Error if do normalization because 0 counts exist in HTO assay; 
```{r}

FGFR4.28HTM.28z_CAR2 <- NormalizeData(FGFR4.28HTM.28z_CAR2, assay = "HTO", normalization.method = "CLR")
FGFR4.28HTM.28z_CAR2 <- MULTIseqDemux(FGFR4.28HTM.28z_CAR2, assay = "HTO", quantile = 0.7, autoThresh = FALSE, maxiter = 5, qrange = seq(from = 0.1, to = 0.9, by = 0.05), verbose = TRUE)

CD276.8HTM.BBz_CAR3 <- NormalizeData(CD276.8HTM.BBz_CAR3, assay = "HTO", normalization.method = "CLR")
CD276.8HTM.BBz_CAR3 <- MULTIseqDemux(CD276.8HTM.BBz_CAR3, assay = "HTO", quantile = 0.7, autoThresh = FALSE, maxiter = 5, qrange = seq(from = 0.1, to = 0.9, by = 0.05), verbose = TRUE)

FGFR4.28HTM.BBz_CD276.8HTM.BBz_BiCisCAR4 <- NormalizeData(FGFR4.28HTM.BBz_CD276.8HTM.BBz_BiCisCAR4, assay = "HTO", normalization.method = "CLR")
FGFR4.28HTM.BBz_CD276.8HTM.BBz_BiCisCAR4 <- MULTIseqDemux(FGFR4.28HTM.BBz_CD276.8HTM.BBz_BiCisCAR4, assay = "HTO", quantile = 0.7, autoThresh = FALSE, maxiter = 5, qrange = seq(from = 0.1, to = 0.9, by = 0.05), verbose = TRUE)

FGFR4.28HTM.28z_CD276.8HTM.BBz_BiCisCAR5 <- NormalizeData(FGFR4.28HTM.28z_CD276.8HTM.BBz_BiCisCAR5, assay = "HTO", normalization.method = "CLR")
FGFR4.28HTM.28z_CD276.8HTM.BBz_BiCisCAR5 <- MULTIseqDemux(FGFR4.28HTM.28z_CD276.8HTM.BBz_BiCisCAR5, assay = "HTO", quantile = 0.7, autoThresh = FALSE, maxiter = 5, qrange = seq(from = 0.1, to = 0.9, by = 0.05), verbose = TRUE)

```

# make a list containing four CARs objects; 
```{r}
CAR2_5.list <- list(FGFR4.28HTM.28z_CAR2, CD276.8HTM.BBz_CAR3, FGFR4.28HTM.BBz_CD276.8HTM.BBz_BiCisCAR4, FGFR4.28HTM.28z_CD276.8HTM.BBz_BiCisCAR5)

for(i in 1: length(CAR2_5.list)){
  print(table(CAR2_5.list[[i]]$MULTI_ID))
  plot = FeatureScatter(CAR2_5.list[[i]], feature1 = "hto_HTO05", feature2 = "hto_HTO06")
  print(plot)
}
```

####using lapply function could get same results as separately HTO demux for each object####
Normalization and demultiplexing by MULTIseqDemux for each object independently; Due to MULTIseqDemux works when HTODemux Error if do normalization because 0 counts exist in HTO assay; 
```{r}
CAR2_5.list <- lapply(X = CAR2_5.list, FUN = function(x) {
    x <- NormalizeData(x, assay = "HTO", normalization.method = "CLR")
    x <- MULTIseqDemux(x, assay = "HTO", quantile = 0.7, autoThresh = FALSE, maxiter = 5, qrange = seq(from = 0.1, to = 0.9, by = 0.05), verbose = TRUE)
})

```
 
# separate negative and singlet from each object 
```{r}
neg.list <- list()
singlets.list <- list()

for(i in 1:length(CAR2_5.list)){
  Idents(CAR2_5.list[[i]]) <- "MULTI_ID"
  neg.list[[i]] <- subset(CAR2_5.list[[i]], idents = "Negative")
  singlets.list[[i]] = subset(CAR2_5.list[[i]], idents = c("HTO05", "HTO06"))
}
  
```

# Add ADT data into each object of negative or singlets list separately
```{r}

neg.list[[1]][["ADT"]] <- CreateAssayObject(counts = FGFR4.28HTM.28z_CAR2.adt[1:17, colnames(x = neg.list[[1]])])
singlets.list[[1]][["ADT"]] <- CreateAssayObject(counts = FGFR4.28HTM.28z_CAR2.adt[1:17,colnames(x = singlets.list[[1]])])

neg.list[[2]][["ADT"]] <- CreateAssayObject(counts = CD276.8HTM.BBz_CAR3.adt[1:17, colnames(x = neg.list[[2]])])
singlets.list[[2]][["ADT"]] <- CreateAssayObject(counts = CD276.8HTM.BBz_CAR3.adt[1:17,colnames(x = singlets.list[[2]])])

neg.list[[3]][["ADT"]] <- CreateAssayObject(counts = FGFR4.28HTM.BBz_CD276.8HTM.BBz_BiCisCAR4.adt[1:17, colnames(x = neg.list[[3]])])
singlets.list[[3]][["ADT"]] <- CreateAssayObject(counts = FGFR4.28HTM.BBz_CD276.8HTM.BBz_BiCisCAR4.adt[1:17,colnames(x = singlets.list[[3]])])


neg.list[[4]][["ADT"]] <- CreateAssayObject(counts = FGFR4.28HTM.28z_CD276.8HTM.BBz_BiCisCAR5.adt[1:17, colnames(x = neg.list[[4]])])
singlets.list[[4]][["ADT"]] <- CreateAssayObject(counts = FGFR4.28HTM.28z_CD276.8HTM.BBz_BiCisCAR5.adt[1:17,colnames(x = singlets.list[[4]])])


```

# get rna or adt matrix of background 
```{r}
neg_rna_mtx <- list()
neg_adt_mtx <- list()

names(neg.list) <- c("FGFR4.28HTM.28z_CAR2.neg", "CD276.8HTM.BBz_CAR3.neg", "FGFR4.28HTM.BBz_CD276.8HTM.BBz_BiCisCAR4.neg", "FGFR4.28HTM.28z_CD276.8HTM.BBz_BiCisCAR5.neg")

for(i in 1: length(neg.list)){
  neg_rna_mtx[[i]] <- as.sparse(GetAssayData(neg.list[[i]], assay = "RNA", slot = 'counts'))
  neg_adt_mtx[[i]] = GetAssayData(neg.list[[i]], assay = "ADT", slot = 'counts') %>% as.matrix()
}

```

# get rna or adt matrix of Positive cells
```{r}
names(singlets.list) <- c("FGFR4.28HTM.28z_CAR2.single", "CD276.8HTM.BBz_CAR3.single", "FGFR4.28HTM.BBz_CD276.8HTM.BBz_BiCisCAR4.single", "FGFR4.28HTM.28z_CD276.8HTM.BBz_BiCisCAR5.single")

singlet_rna_mtx <- list()
positive_adt_mtx <- list()

for(i in 1: length(neg.list)){
  singlet_rna_mtx[[i]] <- as.sparse(GetAssayData(singlets.list[[i]], assay = "RNA", slot = 'counts'))
  positive_adt_mtx[[i]] = GetAssayData(singlets.list[[i]], assay = "ADT", slot = 'counts') %>% as.matrix()
}

```

### removing isotype control background for ADT data and correcting for per-cell technical factor as a covariate seperately
```{r}
isotypes2 = c("M-IgG1", "M-IgG2a","M-IgG2b")
CAR2_dsb.normalized_mtx = DSBNormalizeProtein(cell_protein_matrix = positive_adt_mtx[[1]], empty_drop_matrix = neg_adt_mtx[[1]], use.isotype.control = TRUE, isotype.control.name.vec = isotypes2)

CAR3_dsb.normalized_mtx = DSBNormalizeProtein(cell_protein_matrix = positive_adt_mtx[[2]], empty_drop_matrix = neg_adt_mtx[[2]], use.isotype.control = TRUE, isotype.control.name.vec = isotypes2)

CAR4_dsb.normalized_mtx = DSBNormalizeProtein(cell_protein_matrix = positive_adt_mtx[[3]], empty_drop_matrix = neg_adt_mtx[[3]], use.isotype.control = TRUE, isotype.control.name.vec = isotypes2)

CAR5_dsb.normalized_mtx = DSBNormalizeProtein(cell_protein_matrix = positive_adt_mtx[[4]], empty_drop_matrix = neg_adt_mtx[[4]], use.isotype.control = TRUE, isotype.control.name.vec = isotypes2)

```

# add dsb normalized data to ADT "data" slot
```{r}
CAR3_singlets = singlets.list[[2]] 

CAR3_singlets = SetAssayData(object = CAR3_singlets, assay = "ADT", slot = "data", new.data = CAR3_dsb.normalized_mtx)

```

#####using for loop to calculate dsb.normalized_mtx#####
removing isotype control background for ADT data and correcting for per-cell technical factor as a covariate
```{r}
isotypes2 = c("M-IgG1", "M-IgG2a","M-IgG2b")

dsb.normalized_mtx <- list()

for(i in 1: length(positive_adt_mtx)){
dsb.normalized_mtx[[i]] = DSBNormalizeProtein(cell_protein_matrix = positive_adt_mtx[[i]], empty_drop_matrix = neg_adt_mtx[[i]], use.isotype.control = TRUE, isotype.control.name.vec = isotypes2)
}

```

#### using for loop to add dsb normalized data to ADT "data" slot #####could get same results as seperately analysis
```{r}

for(i in 1: length(singlets.list)){
singlets.list[[i]] = SetAssayData(object = singlets.list[[i]], assay = "ADT", slot = "data", new.data = dsb.normalized_mtx[[i]])
}

```

# QC and filter RNA data; vlnplot or histogram plot QC parameters; 
```{r}
for(i in 1: length(singlets.list)){
  singlets.list[[i]][["percent.mt"]] <- PercentageFeatureSet(singlets.list[[i]], assay = "RNA", pattern = "^MT-")
  print(head(singlets.list[[i]]@meta.data, 5))
  print(mean(singlets.list[[i]]$percent.mt))
  plot1 = VlnPlot(singlets.list[[i]], features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
  plot2 = hist(singlets.list[[i]]$nFeature_RNA)
  plot3 = hist(singlets.list[[i]]$percent.mt)
  print(plot1)
  print(plot2)
  print(plot3)
}

```

# for 4 CAR.objects in singlets.list: filter nFeature_RNA > 500 & nFeature_RNA < 8000 & percent.mt < 20
```{r}

singlets_sub.list <- list()
for(i in 1: length(singlets.list)){
  singlets_sub.list[[i]] <- subset(singlets.list[[i]], subset = nFeature_RNA > 400 & nFeature_RNA < 8000 & percent.mt < 20)
  print(VlnPlot(singlets_sub.list[[i]], features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3))
  print(table(singlets_sub.list[[i]]$MULTI_ID))
}

```

# saveRDS data
```{r}
saveRDS(singlets_sub.list, file = "../SeuratAnalysisData/RDSdata/CAR2_5_singlets_sub.list.rds")

saveRDS(singlets.list, file = "../SeuratAnalysisData/RDSdata/CAR2_5_singlets.list.rds")

```

# Add "MULTI_ID" to object FGFR4.28HTM.BBz_CAR1.sub, in order keep consistent with the replicates of other 4 CARs objects 
```{r}
table(FGFR4.28HTM.BBz_CAR1.sub$orig.ident)
head(FGFR4.28HTM.BBz_CAR1.sub@meta.data)

Idents(FGFR4.28HTM.BBz_CAR1.sub) <- "orig.ident"
levels(Idents(FGFR4.28HTM.BBz_CAR1.sub))

FGFR4.28HTM.BBz_CAR1.sub <-RenameIdents(FGFR4.28HTM.BBz_CAR1.sub, "FGFR4.28HTM.BBz_CAR1_1"="HTO05", "FGFR4.28HTM.BBz_CAR1_2"="HTO06")

level_order <- c("HTO05", "HTO06")

FGFR4.28HTM.BBz_CAR1.sub[["MULTI_ID"]] <- Idents(FGFR4.28HTM.BBz_CAR1.sub)

Idents(FGFR4.28HTM.BBz_CAR1.sub) <- factor(Idents(FGFR4.28HTM.BBz_CAR1.sub), levels = level_order)

head(FGFR4.28HTM.BBz_CAR1.sub@meta.data)
tail(FGFR4.28HTM.BBz_CAR1.sub@meta.data)

```

# Annotation Cells expressing 3A11 to CARpos T cells for FGFR4.28HTM.BBz_CAR1
```{r}
length(WhichCells(FGFR4.28HTM.BBz_CAR1.sub, expression = `3A11` >= 1))

cells.use = WhichCells(FGFR4.28HTM.BBz_CAR1.sub, expression = `3A11` > 0)
cells.use2 = WhichCells(FGFR4.28HTM.BBz_CAR1.sub, expression = `3A11` == 0)

FGFR4.28HTM.BBz_CAR1.sub <- SetIdent(FGFR4.28HTM.BBz_CAR1.sub, cells = cells.use, value = "CARpos T cells")

FGFR4.28HTM.BBz_CAR1.sub <- SetIdent(FGFR4.28HTM.BBz_CAR1.sub, cells = cells.use2, value = "CARneg T cells")

level_order2 <- c("CARpos T cells", "CARneg T cells")

FGFR4.28HTM.BBz_CAR1.sub[["CAR.idents"]] <- Idents(FGFR4.28HTM.BBz_CAR1.sub)

Idents(FGFR4.28HTM.BBz_CAR1.sub) <- factor(Idents(FGFR4.28HTM.BBz_CAR1.sub), levels = level_order2)

head(Idents(FGFR4.28HTM.BBz_CAR1.sub))

tail(Idents(FGFR4.28HTM.BBz_CAR1.sub))

head(FGFR4.28HTM.BBz_CAR1.sub[[]])

```

# check number or percentage of CARpos in two replicates for FGFR4.28HTM.BBz_CAR1.sub
```{r}
table(Idents(FGFR4.28HTM.BBz_CAR1.sub))

table(FGFR4.28HTM.BBz_CAR1.sub$CAR.idents, FGFR4.28HTM.BBz_CAR1.sub$MULTI_ID)

prop.table(table(FGFR4.28HTM.BBz_CAR1.sub$CAR.idents, FGFR4.28HTM.BBz_CAR1.sub$MULTI_ID), margin = 2)

```

# Annotation Cells expressing 3A11 to CARpos T cells for CARs seurat objects in list, first is FGFR4.28HTM.28z_CAR2
```{r}
length(WhichCells(singlets_sub.list[[1]], expression = `3A11` >= 1))

singlets_sub.list[[1]] <- SetIdent(singlets_sub.list[[1]], cells = WhichCells(singlets_sub.list[[1]], expression = `3A11` > 0), value = "CARpos T cells")

singlets_sub.list[[1]] <- SetIdent(singlets_sub.list[[1]], cells = WhichCells(singlets_sub.list[[1]], expression = `3A11` == 0), value = "CARneg T cells")
  
level_order2 <- c("CARpos T cells", "CARneg T cells")

singlets_sub.list[[1]][["CAR.idents"]] <- Idents(singlets_sub.list[[1]])

Idents(singlets_sub.list[[1]]) <- factor(Idents(singlets_sub.list[[1]]), levels = level_order2)

head(singlets_sub.list[[1]][[]])

Idents(singlets_sub.list[[1]]) <- "MULTI_ID"
table(Idents(singlets_sub.list[[1]]))
table(singlets_sub.list[[1]]$CAR.idents, Idents(singlets_sub.list[[1]]))
prop.table(table(singlets_sub.list[[1]]$CAR.idents, Idents(singlets_sub.list[[1]])), margin = 2)

```

# For CD276.8HTM.BBz_CAR3, Annotation Cells expressing MGB7H3LH to CARpos T cells for CARs seurat objects in list
```{r}
length(WhichCells(singlets_sub.list[[2]], expression = `MGB7H3LH` >= 1))

singlets_sub.list[[2]] <- SetIdent(singlets_sub.list[[2]], cells = WhichCells(singlets_sub.list[[2]], expression = `MGB7H3LH` > 0), value = "CARpos T cells")

singlets_sub.list[[2]] <- SetIdent(singlets_sub.list[[2]], cells = WhichCells(singlets_sub.list[[2]], expression = `MGB7H3LH` == 0), value = "CARneg T cells")
  
level_order2 <- c("CARpos T cells", "CARneg T cells")

singlets_sub.list[[2]][["CAR.idents"]] <- Idents(singlets_sub.list[[2]])

Idents(singlets_sub.list[[2]]) <- factor(Idents(singlets_sub.list[[2]]), levels = level_order2)

head(singlets_sub.list[[2]][[]])

Idents(singlets_sub.list[[2]]) <- "MULTI_ID"
table(Idents(singlets_sub.list[[2]]))
table(singlets_sub.list[[2]]$CAR.idents, Idents(singlets_sub.list[[2]]))
prop.table(table(singlets_sub.list[[2]]$CAR.idents, Idents(singlets_sub.list[[2]])), margin = 2)

```

# For FGFR4.28HTM.BBz_CD276.8HTM.BBz_BiCisCAR4, check cell counts of CARpos, CARneg or coexpressing 3A11 and MGB7H3LH 
```{r}
CAR4.coexp = intersect(WhichCells(singlets_sub.list[[3]], expression = `3A11` > 0), WhichCells(singlets_sub.list[[3]], expression = `MGB7H3LH` > 0))
CAR4.pos = union(WhichCells(singlets_sub.list[[3]], expression = `3A11` > 0), WhichCells(singlets_sub.list[[3]], expression = `MGB7H3LH` > 0))
CAR4.neg = intersect(WhichCells(singlets_sub.list[[3]], expression = `3A11` == 0), WhichCells(singlets_sub.list[[3]], expression = `MGB7H3LH` == 0))
                   
length(CAR4.coexp)
length(CAR4.pos)
length(CAR4.neg)
```

# Annotation Cells expressing 3A11 or MGB7H3LH to CARpos T cells for CARs seurat objects in list for FGFR4.28HTM.BBz_CD276.8HTM.BBz_BiCisCAR4
```{r}
singlets_sub.list[[3]] <- SetIdent(singlets_sub.list[[3]], cells = CAR4.pos, value = "CARpos T cells")

singlets_sub.list[[3]] <- SetIdent(singlets_sub.list[[3]], cells = CAR4.neg, value = "CARneg T cells")
  
level_order2 <- c("CARpos T cells", "CARneg T cells")

singlets_sub.list[[3]][["CAR.idents"]] <- Idents(singlets_sub.list[[3]])

Idents(singlets_sub.list[[3]]) <- factor(Idents(singlets_sub.list[[3]]), levels = level_order2)

head(Idents(singlets_sub.list[[3]]))

tail(Idents(singlets_sub.list[[3]]))

head(singlets_sub.list[[3]][[]])

Idents(singlets_sub.list[[3]]) <- "MULTI_ID"
table(Idents(singlets_sub.list[[3]]))
table(singlets_sub.list[[3]]$CAR.idents, Idents(singlets_sub.list[[3]]))
prop.table(table(singlets_sub.list[[3]]$CAR.idents, Idents(singlets_sub.list[[3]])), margin = 2)
```

# For FGFR4.28HTM.28z_CD276.8HTM.BBz_BiCisCAR5, Annotation Cells expressing 3A11 or MGB7H3LH to CARpos T cells
```{r}
CAR5.coexp = intersect(WhichCells(singlets_sub.list[[4]], expression = `3A11` > 0), WhichCells(singlets_sub.list[[4]], expression = `MGB7H3LH` > 0))
CAR5.pos = union(WhichCells(singlets_sub.list[[4]], expression = `3A11` > 0), WhichCells(singlets_sub.list[[4]], expression = `MGB7H3LH` > 0))
CAR5.neg = intersect(WhichCells(singlets_sub.list[[4]], expression = `3A11` == 0), WhichCells(singlets_sub.list[[4]], expression = `MGB7H3LH` == 0))
                   
length(CAR5.coexp)
length(CAR5.pos)
length(CAR5.neg)

singlets_sub.list[[4]] <- SetIdent(singlets_sub.list[[4]], cells = CAR5.pos, value = "CARpos T cells")

singlets_sub.list[[4]] <- SetIdent(singlets_sub.list[[4]], cells = CAR5.neg, value = "CARneg T cells")
  
level_order2 <- c("CARpos T cells", "CARneg T cells")

singlets_sub.list[[4]][["CAR.idents"]] <- Idents(singlets_sub.list[[4]])

Idents(singlets_sub.list[[4]]) <- factor(Idents(singlets_sub.list[[4]]), levels = level_order2)

head(singlets_sub.list[[4]][[]])

Idents(singlets_sub.list[[4]]) <- "MULTI_ID"

table(Idents(singlets_sub.list[[4]]))

table(singlets_sub.list[[4]]$CAR.idents, Idents(singlets_sub.list[[4]]))

prop.table(table(singlets_sub.list[[4]]$CAR.idents, Idents(singlets_sub.list[[4]])), margin = 2)

```

# make a list containing all 5 CAR seurat objects
```{r}
all.CARs.list <- list(FGFR4.28HTM.BBz_CAR1.sub, singlets_sub.list[[1]], singlets_sub.list[[2]], singlets_sub.list[[3]], singlets_sub.list[[4]])

all.CARs.list[[1]]$Group <- "FGFR4.28HTM.BBz_CAR1"
all.CARs.list[[2]]$Group <- "FGFR4.28HTM.28z_CAR2"
all.CARs.list[[3]]$Group <- "CD276.8HTM.BBz_CAR3"
all.CARs.list[[4]]$Group <- "FGFR4.28HTM.BBz_CD276.8HTM.BBz_BiCisCAR4"
all.CARs.list[[5]]$Group <- "FGFR4.28HTM.28z_CD276.8HTM.BBz_BiCisCAR5"

head(all.CARs.list[[1]][[]])
head(all.CARs.list[[5]][[]])

```

# combine all 5 CAR objects into one big objects
```{r}
all.CARs.combined <- merge(FGFR4.28HTM.BBz_CAR1.sub, y = c(singlets_sub.list[[1]], singlets_sub.list[[2]], singlets_sub.list[[3]], singlets_sub.list[[4]]), project = "all.CARs.combined", merge.data = TRUE)

GetAssayData(all.CARs.combined)[1:10, 1:15]

DefaultAssay(all.CARs.combined) <- "ADT"

GetAssayData(all.CARs.combined)[1:10, 1:15]

head(all.CARs.combined@meta.data)
tail(all.CARs.combined@meta.data)

all.CARs.combined
head(colnames(all.CARs.combined))
tail(colnames(all.CARs.combined))

table(all.CARs.combined$orig.ident)

```

install glmGamPoi
```{r}
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("glmGamPoi")
options(download.file.method = "wininet")

library(glmGamPoi)
```

#### use sct to remove lane effects, regress out orig.ident or percent.mt
```{r}

all.CARs.combined <- SCTransform(all.CARs.combined, method = "glmGamPoi", vars.to.regress = c("orig.ident", "percent.mt"), variable.features.n = 5000)

saveRDS(all.CARs.combined, file = "all.CARs.combined_sct.rds")
```

## Quality check
```{r}
VlnPlot(all.CARs.combined, features = c("nFeature_SCT", "nCount_SCT"), ncol = 2, pt.size =0.1, group.by = "orig.ident")
```

# use sct for wnn
```{r}
DefaultAssay(all.CARs.combined) <- 'SCT'
all.CARs.combined <- RunPCA(all.CARs.combined,  assay="SCT", reduction.name = "pca_sct", reduction.key = "pca_sct_")
ElbowPlot(all.CARs.combined, reduction = "pca_sct", ndims = 50)

```

# normalize and scaling, runPCA for ADT data
```{r}
DefaultAssay(all.CARs.combined) <- 'ADT'
all.CARs.combined <- NormalizeData(all.CARs.combined, normalization.method = 'CLR', margin = 2)
all.CARs.combined <- ScaleData(all.CARs.combined, vars.to.regress = "orig.ident") %>% RunPCA(reduction.name = 'apca')

ElbowPlot(all.CARs.combined, reduction = "apca", ndims = 13)
```

## check dims no. for each modality
```{r}
all.CARs.combined[["pca_sct"]]
all.CARs.combined[["apca"]]
```

# Construct weighted nearest neighbor(WNN) graph
```{r}
all.CARs.combined <- FindMultiModalNeighbors(
  all.CARs.combined, reduction.list = list("pca_sct", "apca"), 
  dims.list = list(1:50, 1:13), modality.weight.name = c("sctRNA.weight","clrADT.weight")
)
```

## run wnnUMAP, find Clusters using resolution 0.4                   
```{r}
all.CARs.combined <- RunUMAP(all.CARs.combined, nn.name = "weighted.nn", reduction.name = "wnn.umap", reduction.key = "wnnUMAP_")

all.CARs.combined <- FindClusters(all.CARs.combined, graph.name = "wsnn", algorithm = 3, resolution = 0.4, verbose = FALSE)

saveRDS(all.CARs.combined, "all.CARs.combined_sct_wnn0.4.rds")
```

##plot for all.CARs.combined, check the effect after correct batch effect
```{r}
DimPlot(all.CARs.combined, reduction = 'wnn.umap', label = TRUE, group.by = "seurat_clusters")
DimPlot(all.CARs.combined, reduction = 'wnn.umap', label = FALSE, group.by = "Group")
DimPlot(all.CARs.combined, reduction = 'wnn.umap', label = TRUE, split.by = "orig.ident")
FeaturePlot(all.CARs.combined, reduction = "wnn.umap",features = c("CD4","CD8A"), min.cutoff = "q05", max.cutoff = "q95")
```
#
```{r}
FeaturePlot(all.CARs.combined, reduction = "wnn.umap",features = c("adt_CD4", "adt_CD8a"), min.cutoff = "q05", max.cutoff = "q95")
DimPlot(all.CARs.combined, reduction = 'wnn.umap', label = FALSE, group.by = "Group")
```
# plot group by CART group for all.CARs.combined2
```{r}
level_order <- c("FGFR4.CD28HTM.BBz_CAR1", "FGFR4.CD28HTM.CD28z_CAR2", "CD276.CD8HTM.BBz_CAR3", "CD28HTM.BBz_CD8HTM.BBz_BiCisCAR4", "CD28HTM.CD28z_CD8HTM.BBz_BiCisCAR5" )

all.CARs.combined$Group <- factor(all.CARs.combined$Group, levels = level_order)
DimPlot(all.CARs.combined, reduction = 'wnn.umap', group.by = 'seurat_clusters', split.by = "Group", label = TRUE, repel = TRUE, label.size = 5)+ ggtitle("WNN")
```

# chech batch effect in all.CARs.combined
```{r}
level_order2 <- c("FGFR4.CD28HTM.BBz_CAR1_HTO05", "FGFR4.CD28HTM.BBz_CAR1_HTO06","FGFR4.CD28HTM.CD28z_CAR2_HTO05", "FGFR4.CD28HTM.CD28z_CAR2_HTO06", "CD276.CD8HTM.BBz_CAR3_HTO05", "CD276.CD8HTM.BBz_CAR3_HTO06", "CD28HTM.BBz_CD8HTM.BBz_BiCisCAR4_HTO05", "CD28HTM.BBz_CD8HTM.BBz_BiCisCAR4_HTO06", "CD28HTM.CD28z_CD8HTM.BBz_BiCisCAR5_HTO05", "CD28HTM.CD28z_CD8HTM.BBz_BiCisCAR5_HTO06")

all.CARs.combined$Group.IDs <- paste(all.CARs.combined$Group, all.CARs.combined$MULTI_ID, sep = "_")

head(all.CARs.combined[[]])

all.CARs.combined$Group.IDs <- factor(all.CARs.combined$Group.IDs, levels = level_order2)
DimPlot(all.CARs.combined, reduction = 'wnn.umap', group.by = 'seurat_clusters', split.by = "Group.IDs")+ ggtitle("check batch effect")
```

###check CARpos vs CARneg, split by Group.CARidents
```{r}
all.CARs.combined$Group.CARidents <- paste(all.CARs.combined$Group, all.CARs.combined$CAR.idents, sep = "_")

head(all.CARs.combined[[]])

level_order3 <- c("FGFR4.CD28HTM.BBz_CAR1_CARneg T cells", "FGFR4.CD28HTM.BBz_CAR1_CARpos T cells","FGFR4.CD28HTM.CD28z_CAR2_CARneg T cells", "FGFR4.CD28HTM.CD28z_CAR2_CARpos T cells", "CD276.CD8HTM.BBz_CAR3_CARneg T cells", "CD276.CD8HTM.BBz_CAR3_CARpos T cells", "CD28HTM.BBz_CD8HTM.BBz_BiCisCAR4_CARneg T cells", "CD28HTM.BBz_CD8HTM.BBz_BiCisCAR4_CARpos T cells", "CD28HTM.CD28z_CD8HTM.BBz_BiCisCAR5_CARneg T cells", "CD28HTM.CD28z_CD8HTM.BBz_BiCisCAR5_CARpos T cells")

all.CARs.combined$Group.CARidents <- factor(all.CARs.combined$Group.CARidents, levels = level_order3)
DimPlot(all.CARs.combined, reduction = 'wnn.umap', group.by = 'seurat_clusters', split.by = "Group.CARidents")+ ggtitle("wnn0.4 clusters for CARpos vs CARneg")
```

##plot rna or ADT Features expression on each cluster of all.CARs.combined based on clustering resolution 0.4
```{r}
DefaultAssay(all.CARs.combined) <- "SCT"
rna.features <- c("CD3E", "CD4", "CD8A", "TRDC", "TRGC2","CD27", "FAS","SELL", "IL7R","IL2RA", "TNFRSF9", "LAG3", "ENTPD1", "PDCD1", "HAVCR2", "TOP2A","MKI67", "PCNA", "MCM5","KLRG1", "GZMK", "GZMA", "GNLY", "NKG7", "PRF1", "IFNG", "ITGAE", "CD69")
adt.features <- c("adt_CD3", "adt_CD4", "adt_CD8a", "adt_CD45RA", "adt_CD45RO", 
                  "adt_CD27","adt_CD95", "adt_CD62L", "adt_CD25", "adt_CD137",
                  "adt_LAG-3", "adt_CD39", "adt_PD-1", "adt_TIM-3")
VlnPlot(all.CARs.combined, features = rna.features, split.by = "seurat_clusters", pt.size = 0, stack=T, flip = T) + NoLegend()

VlnPlot(all.CARs.combined, features = adt.features, split.by = "seurat_clusters", pt.size = 0, stack=T, flip = T)  + NoLegend()
```

##FeaturePlot to show the RNA and ADT data for specific markers
```{r}
##RNA FeaturePlot
FeaturePlot(all.CARs.combined, reduction = "wnn.umap", features = c("CD3E","CD4", "CD8A", "CD27", "FAS","SELL", "IL2RA", "TNFRSF9", "LAG3", "ENTPD1", "PDCD1", "HAVCR2"), ncol =4, min.cutoff = "q05", max.cutoff = "q95")

##ADT FeaturePlot
FeaturePlot(all.CARs.combined, reduction = "wnn.umap", features = c("adt_CD3", "adt_CD4", "adt_CD8a", "adt_CD27", "adt_CD95","adt_CD62L", "adt_CD25","adt_CD137", "adt_LAG-3", "adt_CD39", "adt_PD-1", "adt_TIM-3", "adt_CD45RA", "adt_CD45RO"), cols = c("grey", "red"), ncol = 4, min.cutoff = "q05", max.cutoff = "q95")

```

#RNA and ADT pairwise FeaturePlot, save paired RNA and ADT featureplot in one PDF
```{r}
p1 <- FeaturePlot(all.CARs.combined, reduction = "wnn.umap", features = c("CD3E","CD4", "CD8A", "CD27"), ncol = 4, min.cutoff = "q05", max.cutoff = "q95")

p2 <- FeaturePlot(all.CARs.combined, reduction = "wnn.umap", features = c("adt_CD3", "adt_CD4", "adt_CD8a","adt_CD27"), cols = c("grey", "red"), ncol = 4, min.cutoff = "q05", max.cutoff = "q95")

##par(mfrow=c(2,1))##save two plots in one PDF with two page
pdf("../SeuratAnalysisData/Data/SCT_wnn0.4_ClustersAnnotation&Percentages/FeaturePlot_rna&adt1.pdf", width = 15.6, height = 7)
  p1 / p2
  dev.off()
```

```{r}
p3 <- FeaturePlot(all.CARs.combined2, reduction = "wnn.umap", features = c("FAS","SELL", "IL2RA", "TNFRSF9"), ncol = 4, min.cutoff = "q05", max.cutoff = "q95")

p4 <- FeaturePlot(all.CARs.combined2, reduction = "wnn.umap", features = c("adt_CD95","adt_CD62L", "adt_CD25","adt_CD137"), cols = c("grey", "red"), ncol = 4, min.cutoff = "q05", max.cutoff = "q95")

pdf("../SeuratAnalysisData/Data/SCT_wnn0.4_ClustersAnnotation&Percentages/FeaturePlot_rna&adt2.pdf", width = 15.6, height = 7)
  p3 / p4
  dev.off()
```

```{r}
p5 <- FeaturePlot(all.CARs.combined2, reduction = "wnn.umap", features = c("LAG3", "ENTPD1", "PDCD1", "HAVCR2"), ncol = 4, min.cutoff = "q05", max.cutoff = "q95")

p6 <- FeaturePlot(all.CARs.combined2, reduction = "wnn.umap", features = c("adt_LAG-3", "adt_CD39", "adt_PD-1", "adt_TIM-3"), cols = c("grey", "red"), ncol = 4, min.cutoff = "q05", max.cutoff = "q95")

pdf("../SeuratAnalysisData/Data/SCT_wnn0.4_ClustersAnnotation&Percentages/FeaturePlot_rna&adt3.pdf", width = 15.6, height = 7)
  p5 / p6
  dev.off()
```

```{r}
p7 <- FeaturePlot(all.CARs.combined2, reduction = "wnn.umap", features = c("adt_CD45RA", "adt_CD45RO"), ncol =2, cols = c("grey", "red"), min.cutoff = "q05", max.cutoff = "q95")

pdf("../SeuratAnalysisData/Data/SCT_wnn0.4_ClustersAnnotation&Percentages/FeaturePlot_adt4.pdf", width = 7.8, height = 3.5)
  p7
  dev.off()
```

# Find conserved markers for cluster annotation
# find rna markers for every cluster compared to all remaining cells, report only the positive ones
```{r}
Idents(all.CARs.combined) <- "seurat_clusters"

allCARs.markers <- FindAllMarkers(all.CARs.combined, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)
allCARs.markers %>%
    group_by(cluster) %>%
    slice_max(n = 2, order_by = avg_log2FC)
```

## DoHeatmap() generates an expression heatmap for given cells and features. In this case, we are plotting the top 20 markers (or all markers if less than 20) for each cluster.
```{r}
allCARs.markers %>%
    group_by(cluster) %>%
    top_n(n = 20, wt = avg_log2FC) -> top20

DefaultAssay(all.CARs.combined) <- "SCT"

DoHeatmap(all.CARs.combined, features = top20$gene) + scale_fill_gradientn(colours = rev(brewer.pal(11 ,"RdBu"))) + NoLegend()

write.table(allCARs.markers, file = "../SeuratAnalysisData/Data/allCARs.combined2_sct_wnn0.4_allclusters_conserved_markers.tsv", sep ="\t", quote = FALSE, row.names = FALSE, col.names = TRUE)

write.table(top20, file = "../SeuratAnalysisData/Data/allCARs.combined2_sct_wnn0.4_conserved_markers_top20.tsv", sep ="\t", quote = FALSE, row.names = FALSE, col.names = TRUE)
```

## Differential expression analysis: Identify genes specificly expressed in some cluster####
## Find differential genes expressed in cluster4_CD8 compared with other CD8+T
```{r}
DefaultAssay(all.CARs.combined) <- "SCT"
cluster4.markers <- FindMarkers(all.CARs.combined,
                          ident.1 = 4,
                          ident.2 = c(0,2,3,5,6,7,10,12), only.pos = TRUE, logfc.threshold = 0.25)
```

##order by avg_log2Fc and save
```{r}
cluster4.markers <- cluster4.markers %>% arrange(desc(avg_log2FC))
cluster4.markers$gene <- rownames(cluster4.markers)
cluster4.markers <- cluster4.markers[, c(6,2,1,3,4,5)]
top25_C4 <- top_n(cluster4.markers, n=25, wt = avg_log2FC)
write.table(cluster4.markers, file = "../SeuratAnalysisData/Data/allCARs.combined2_sct_wnn0.4_cluster4vsotherCD8t_markers.tsv", sep ="\t", quote = FALSE, row.names = FALSE, col.names = TRUE)
```

##Annotation clusters based on sct markers and ADT data
```{r}
Idents(all.CARs.combined) <- "seurat_clusters"
all.CARs.combined <- RenameIdents(all.CARs.combined, `0` = "C0_CD4+ Activated T(TNFRSF4+)", `1` = "C1_CD8+ Early Tem(GZMK+CD27+)", `2` = "C2_CD8 Tex(CD39+CD103+)", `3` = "C3_CD8+ Late Tem(GZMA+CD27-)", `4` = "C4_T(ARHGAP15+)", `5` = "C5_CD8+ T Cycling/G2M", `6` = "C6_Activated Teff(IFNG+)", `7` = "C7_CD8+ Teff(GNLY+GZMB+)", `8` = "C8_CD8+ MThi T(CD69+JUN+)", `9` = "C9_CD4+ Tscm(KLF2+TCF7+)", `10` = "C10_CD8+ Tcm(KLF2+SELL+)", `11` = "C11_γδ T cell", `12` = "C12_CD4+ Activated T(TRBV20-1)", `13` = "C13_CD4+ Activated T(TRBV5-1)", `14` = "C14_CD8+ Temra(TRBV9)", `15` = "C15_CD4+ Activated T(TRBV7-2)", `16` = "C16_CD8+ Tem(TRBV7-9)", `17` = "C17_CD8+ Tem(TRBV30)", `18` = "C18_NKT(FCER1G+TRDC+)", `19` = "C19_CD4+ Activated T(TRBV6-5)", `20` = "C20_LYZ+")

level_order <- c("C0_CD4+ Activated T(TNFRSF4+)", "C1_CD8+ Early Tem(GZMK+CD27+)", "C2_CD8 Tex(CD39+CD103+)", "C3_CD8+ Late Tem(GZMA+CD27-)", "C4_T(ARHGAP15+)", "C5_CD8+ T Cycling/G2M", "C6_Activated Teff(IFNG+)", "C7_CD8+ Teff(GNLY+GZMB+)", "C8_CD8+ MThi T(CD69+JUN+)", "C9_CD4+ Tscm(KLF2+TCF7+)", "C10_CD8+ Tcm(KLF2+SELL+)", "C11_γδ T cell", "C12_CD4+ Activated T(TRBV20-1)", "C13_CD4+ Activated T(TRBV5-1)", "C14_CD8+ Temra(TRBV9)", "C15_CD4+ Activated T(TRBV7-2)", "C16_CD8+ Tem(TRBV7-9)", "C17_CD8+ Tem(TRBV30)",  "C18_NKT(FCER1G+TRDC+)", "C19_CD4+ Activated T(TRBV6-5)", "C20_LYZ+")

all.CARs.combined[["celltype0.4"]] <- Idents(all.CARs.combined)
Idents(all.CARs.combined) <- factor(Idents(all.CARs.combined), levels = level_order)

p8 <- DimPlot(all.CARs.combined, reduction = 'wnn.umap', group.by = 'seurat_clusters', label = TRUE, repel = TRUE, label.size = 3) + NoLegend() + ggtitle("SCT_wnnUMAP")

p9 <- DimPlot(all.CARs.combined, reduction = 'wnn.umap', group.by = 'celltype0.4', label = TRUE, repel = TRUE, label.size = 3) + NoLegend() + ggtitle("Annotation")
p8+p9
```

## combined cells from C12,C13,C15,C19 using specific TCR with C0_CD4+ Activated T, 
## combined cells from C16,C17 using specific TCR with C1_CD8+ early Tem, save idents as population 
```{r}
Idents(all.CARs.combined2) <- "seurat_clusters"
all.CARs.combined <- RenameIdents(all.CARs.combined, `0` = "C0_CD4+ Activated T", `1` = "C1_CD8+ Early Tem(GZMK+CD27+)", `2` = "C2_CD8 Tex(CD39+CD103+)", `3` = "C3_CD8+ Late Tem(GZMA+CD27-)", `4` = "C4_T(ARHGAP15+)", `5` = "C5_CD8+ T Cycling/G2M", `6` = "C6_Activated Teff(IFNG+)", `7` = "C7_CD8+ Teff(GNLY+GZMB+)", `8` = "C8_CD8+ MThi T(CD69+JUN+)", `9` = "C9_CD4+ Tscm(KLF2+TCF7+)", `10` = "C10_CD8+ Tcm(KLF2+SELL+)", `11` = "C11_γδ T cell", `12` = "C0_CD4+ Activated T", `13` = "C0_CD4+ Activated T", `14` = "C14_CD8+ Temra(TRBV9)", `15` = "C0_CD4+ Activated T", `16` = "C1_CD8+ Early Tem(GZMK+CD27+)", `17` = "C1_CD8+ Early Tem(GZMK+CD27+)", `18` = "C18_NKT(FCER1G+TRDC+)", `19` = "C0_CD4+ Activated T", `20` = "C20_LYZ+")


population_level_order <- c("C0_CD4+ Activated T", "C9_CD4+ Tscm(KLF2+TCF7+)", "C4_T(ARHGAP15+)", "C6_Activated Teff(IFNG+)", "C1_CD8+ Early Tem(GZMK+CD27+)", "C2_CD8 Tex(CD39+CD103+)", "C3_CD8+ Late Tem(GZMA+CD27-)", "C5_CD8+ T Cycling/G2M",  "C7_CD8+ Teff(GNLY+GZMB+)", "C8_CD8+ MThi T(CD69+JUN+)", "C10_CD8+ Tcm(KLF2+SELL+)", "C14_CD8+ Temra(TRBV9)", "C11_γδ T cell", "C18_NKT(FCER1G+TRDC+)", "C20_LYZ+")

all.CARs.combined[["population"]] <- Idents(all.CARs.combined)
Idents(all.CARs.combined) <- factor(Idents(all.CARs.combined), levels = population_level_order)

head(all.CARs.combined[[]])
levels(Idents(all.CARs.combined))

p10 <- DimPlot(all.CARs.combined, reduction = 'wnn.umap', group.by = 'seurat_clusters', label = TRUE, repel = TRUE, label.size = 3)+ ggtitle("SCT_wnn0.4") + NoLegend()

p11 <- DimPlot(all.CARs.combined, reduction = 'wnn.umap', group.by = 'population', label = TRUE, repel = TRUE, label.size = 3)+ ggtitle("Cell populations for SCT_wnn0.4")+ NoLegend()

p10 + p11
```

##plot cell counts per Group.IDs
```{r}
table1 <- as.data.frame(table(Idents(all.CARs.combined), all.CARs.combined$Group.IDs))

colnames(table1) <- c("cellpopulation", "Group_Replicates", "cellcounts")

write.table(table1, file = paste0("../SeuratAnalysisData/Data/", "cellcounts_Per_Population_CARgroups_sctwnn0.4.tsv"), sep ="\t", quote = FALSE, row.names = FALSE, col.names = TRUE )

```

##plot cell counts per Group.CAR.idents
```{r}
table2 <- as.data.frame(table(Idents(all.CARs.combined), all.CARs.combined$Group.CARidents))

colnames(table2) <- c("cellpopulation", "Group_CARidents", "cellcounts")

write.table(table2, file = paste0("../SeuratAnalysisData/Data/", "cellcounts_Per_Population_CARidents_Per_groups_sctwnn0.4.tsv"), sep ="\t", quote = FALSE, row.names = FALSE, col.names = TRUE )

```

#design colors for each cell population 
```{r}
library(RColorBrewer)
display.brewer.all()

brewer.pal(8,"Dark2")##display color name

show_col(brewer.pal(8,"Dark2"))
```

##to display cols
```{r}
library(scales)

cols = c("#00AFBB", "#E7B800", "#A6D854", "#A6761D", "#66A61E", "#FF7F00", "#9970AB", "#D53E4F", "#2171B5","#F781BF", "#FE9929", "#6BAED6", "#FC4E2A", "#CAB2D6", "#666666")
show_col(cols)

```
# using new cols to plot wnnUMAP
```{r}
population_cols <- c("C0_CD4+ Activated T" = "#00AFBB", "C9_CD4+ Tscm(KLF2+TCF7+)"="#E7B800", "C4_T(ARHGAP15+)"="#A6D854", "C6_Activated Teff(IFNG+)"="#A6761D", "C1_CD8+ Early Tem(GZMK+CD27+)"= "#66A61E", "C2_CD8 Tex(CD39+CD103+)"="#FF7F00", "C3_CD8+ Late Tem(GZMA+CD27-)"="#9970AB", "C5_CD8+ T Cycling/G2M"="#D53E4F",  "C7_CD8+ Teff(GNLY+GZMB+)"="#2171B5", "C8_CD8+ MThi T(CD69+JUN+)"="#F781BF", "C10_CD8+ Tcm(KLF2+SELL+)"="#FE9929", "C14_CD8+ Temra(TRBV9)"="#6BAED6", "C11_γδ T cell"="#FC4E2A", "C18_NKT(FCER1G+TRDC+)"="#CAB2D6", "C20_LYZ+"="#666666")

p12 <- DimPlot(all.CARs.combined, reduction = 'wnn.umap', group.by = 'population', cols =population_cols, label = TRUE, repel = TRUE, label.size = 3)+ ggtitle("Cell populations for SCT_wnn0.4")+ NoLegend()

p13 <- DimPlot(all.CARs.combined, reduction = 'wnn.umap', group.by = 'population', cols =population_cols, label = FALSE)+ ggtitle("Cell populations for SCT_wnn0.4")

p12 + p13
```

# plot overlap view of 5 CAR groups, using defined cols
```{r}
CARgroups_cols <- c("FGFR4.CD28HTM.BBz_CAR1" = "#FF6699", "FGFR4.CD28HTM.CD28z_CAR2"="#00CCCC", "CD276.CD8HTM.BBz_CAR3"="#9966FF",  "CD28HTM.BBz_CD8HTM.BBz_BiCisCAR4"="#CC6633", "CD28HTM.CD28z_CD8HTM.BBz_BiCisCAR5"="#0066FF")

DimPlot(all.CARs.combined, reduction = 'wnn.umap', group.by = 'Group', cols = CARgroups_cols, label = FALSE)+ ggtitle("batch effect check by CAR groups")
```

##Split.by CAR groups or replicates
```{r}
DimPlot(all.CARs.combined, reduction = 'wnn.umap', group.by = 'population', split.by = "Group", cols =population_cols)+ ggtitle("wnn0.4 cell populations annotation for each CAR")

t1 <- as.data.frame(table(all.CARs.combined$Group))
```

#DimPlot split.by Group and replicate IDs, and list cell counts
## two rows and 5 columns to show the Group.IDs dimplot
```{r}
level_order2 <- c("FGFR4.CD28HTM.BBz_CAR1_HTO05","FGFR4.CD28HTM.CD28z_CAR2_HTO05","CD276.CD8HTM.BBz_CAR3_HTO05", "CD28HTM.BBz_CD8HTM.BBz_BiCisCAR4_HTO05", "CD28HTM.CD28z_CD8HTM.BBz_BiCisCAR5_HTO05", "FGFR4.CD28HTM.BBz_CAR1_HTO06", "FGFR4.CD28HTM.CD28z_CAR2_HTO06","CD276.CD8HTM.BBz_CAR3_HTO06",  "CD28HTM.BBz_CD8HTM.BBz_BiCisCAR4_HTO06", "CD28HTM.CD28z_CD8HTM.BBz_BiCisCAR5_HTO06")

all.CARs.combined$Group.IDs <- factor(all.CARs.combined$Group.IDs, levels = level_order2)

population_level_order <- c("C0_CD4+ Activated T", "C9_CD4+ Tscm(KLF2+TCF7+)", "C4_T(ARHGAP15+)", "C6_Activated Teff(IFNG+)", "C2_CD8 Tex(CD39+CD103+)", "C5_CD8+ T Cycling/G2M","C8_CD8+ MThi T(CD69+JUN+)", "C1_CD8+ Early Tem(GZMK+CD27+)", "C3_CD8+ Late Tem(GZMA+CD27-)", "C7_CD8+ Teff(GNLY+GZMB+)", "C10_CD8+ Tcm(KLF2+SELL+)", "C14_CD8+ Temra(TRBV9)", "C11_γδ T cell", "C18_NKT(FCER1G+TRDC+)", "C20_LYZ+")

population_cols <- c("C0_CD4+ Activated T" = "#00AFBB", "C9_CD4+ Tscm(KLF2+TCF7+)"="#E7B800", "C4_T(ARHGAP15+)"="#A6D854", "C6_Activated Teff(IFNG+)"="#A6761D", "C2_CD8 Tex(CD39+CD103+)"="#FF7F00", "C5_CD8+ T Cycling/G2M"="#D53E4F", "C8_CD8+ MThi T(CD69+JUN+)"="#F781BF", "C1_CD8+ Early Tem(GZMK+CD27+)"= "#66A61E", "C3_CD8+ Late Tem(GZMA+CD27-)"="#9970AB", "C7_CD8+ Teff(GNLY+GZMB+)"="#2171B5", "C10_CD8+ Tcm(KLF2+SELL+)"="#FE9929", "C14_CD8+ Temra(TRBV9)"="#6BAED6", "C11_γδ T cell"="#FC4E2A", "C18_NKT(FCER1G+TRDC+)"="#CAB2D6", "C20_LYZ+"="#666666")

Idents(all.CARs.combined) <- factor(Idents(all.CARs.combined), levels = population_level_order)

DimPlot(all.CARs.combined, reduction = 'wnn.umap', group.by = 'population', split.by = "Group.IDs", cols =population_cols, ncol=5)+ ggtitle("wnn0.4 cell populations annotation for each CAR replicate") #5cols*2rows, 12w*22h PDF 14.6*9.07

t2 <- as.data.frame(table(all.CARs.combined$Group.IDs))
```

#DimPlot split.by Group and CARidents, and list cell counts
```{r}
DimPlot(all.CARs.combined, reduction = 'wnn.umap', group.by = 'population', split.by = "Group.CARidents", cols =population_cols, ncol=2)+ ggtitle("wnn0.4 cell populations annotation for CARpos vs CARneg") #2cols*5rows, 12w*22h PDF; 14.6*9.07

t3 <- as.data.frame(table(all.CARs.combined2$Group.CARidents))
```

##Add Idents Group.IDs_CARidents 
```{r}
all.CARs.combined$Group.IDs_CARidents <- paste(all.CARs.combined$Group.IDs, all.CARs.combined$CAR.idents, sep = "_")

head(all.CARs.combined[[]])

level_order4 <- c("FGFR4.CD28HTM.BBz_CAR1_HTO05_CARneg T cells","FGFR4.CD28HTM.BBz_CAR1_HTO06_CARneg T cells", "FGFR4.CD28HTM.BBz_CAR1_HTO05_CARpos T cells","FGFR4.CD28HTM.BBz_CAR1_HTO06_CARpos T cells", "FGFR4.CD28HTM.CD28z_CAR2_HTO05_CARneg T cells", "FGFR4.CD28HTM.CD28z_CAR2_HTO06_CARneg T cells", "FGFR4.CD28HTM.CD28z_CAR2_HTO05_CARpos T cells", "FGFR4.CD28HTM.CD28z_CAR2_HTO06_CARpos T cells", "CD276.CD8HTM.BBz_CAR3_HTO05_CARneg T cells", "CD276.CD8HTM.BBz_CAR3_HTO06_CARneg T cells", "CD276.CD8HTM.BBz_CAR3_HTO05_CARpos T cells", "CD276.CD8HTM.BBz_CAR3_HTO06_CARpos T cells", "CD28HTM.BBz_CD8HTM.BBz_BiCisCAR4_HTO05_CARneg T cells", "CD28HTM.BBz_CD8HTM.BBz_BiCisCAR4_HTO06_CARneg T cells", "CD28HTM.BBz_CD8HTM.BBz_BiCisCAR4_HTO05_CARpos T cells", "CD28HTM.BBz_CD8HTM.BBz_BiCisCAR4_HTO06_CARpos T cells", "CD28HTM.CD28z_CD8HTM.BBz_BiCisCAR5_HTO05_CARneg T cells", "CD28HTM.CD28z_CD8HTM.BBz_BiCisCAR5_HTO06_CARneg T cells", "CD28HTM.CD28z_CD8HTM.BBz_BiCisCAR5_HTO05_CARpos T cells", "CD28HTM.CD28z_CD8HTM.BBz_BiCisCAR5_HTO06_CARpos T cells")

all.CARs.combined$Group.IDs_CARidents <- factor(all.CARs.combined$Group.IDs_CARidents, levels = level_order4)

levels(all.CARs.combined$Group.IDs_CARidents)
```

#DimPlot split.by Group, replicates and CARidents, and list cell counts
```{r}
DimPlot(all.CARs.combined, reduction = 'wnn.umap', group.by = 'population', split.by = "Group.IDs_CARidents", cols =population_cols, ncol=4)+ ggtitle("wnn0.4 cell populations annotation for CAR groups per relicate and CARpos vs CARneg") #4cols*5rows, 19.8w*22h PDF; 14.6h*13.8w

t4 <- as.data.frame(table(all.CARs.combined2$Group.IDs_CARidents))
```

##split Seurat objects based on Groups for plotting cluster percentage for each CAR group
```{r}
DefaultAssay(all.CARs.combined) <- "SCT"
Idents(all.CARs.combined) <- "Group"
all.CARs.list <- SplitObject(all.CARs.combined, split.by = "Group")
```

##using For loop to do bar plots of each population percentage in CAR group, split by Replicates
```{r}
for(i in 1:length(all.CARs.list)){
  Idents(all.CARs.list[[i]]) <- "population"
  df_i <- as.data.frame(prop.table(table(Idents(all.CARs.list[[i]]), all.CARs.list[[i]]$MULTI_ID), margin = 2))
  colnames(df_i) <- c("population", "Replicate","percentage")
  df_i$Freq <- df_i[,3]*100
  
  plot = ggplot(df_i, aes(x=factor(population, level = population_level_order), y=Freq, fill= Replicate)) + 
    geom_bar(stat = "identity", width = 0.5, position = position_dodge(0.8)) + 
    ggtitle(paste(names(all.CARs.list)[i], "clusters distribution", sep = "_")) + ylab("% in TIL") +
    theme_classic() +
    theme(axis.title = element_text( face = "bold", size=10),
          axis.line = element_line(size = 0.5),
          axis.title.x = element_blank(),
          axis.text.x = element_text(size = 8, angle=45, hjust = 1),
          axis.text.y = element_text(size = 8)) + 
    scale_y_continuous(expand = expansion(mult = c(0, .05)))
 
  write.table(df_i, file = paste0("../SeuratAnalysisData/Data/", "df_", names(all.CARs.list)[i], "PopulationPercentages_replicates_sct-wnn0.4.tsv"), sep ="\t", quote = FALSE, row.names = FALSE, col.names = TRUE )
  
  assign(paste0("df_", names(all.CARs.list)[i]), df_i)
  
  print(plot)
}
```

## Plot cell counts per each population for CARpos vs CARneg and each replicates
```{r}
 for(i in 1:length(all.CARs.list)){
  Idents(all.CARs.list[[i]]) <- "population"
  df_i <- as.data.frame(table(Idents(all.CARs.list[[i]]), all.CARs.list[[i]]$CAR.idents, all.CARs.list[[i]]$MULTI_ID))
  colnames(df_i) <- c("population", "CAR.idents", "Replicates", "cellcounts")
  
  write.table(df_i, file = paste0("df_", names(all.CARs.list)[i], "cellcounts_Per_Population_CARposVSCARneg_sctwnn0.4.tsv"), sep ="\t", quote = FALSE, row.names = FALSE, col.names = TRUE )
  }
  write.table(df_i, file = paste0("../Data/", "df_", names(all.CARs.list)[i], "cellcounts_Per_Population_CARposVSCARneg_sctwnn0.4.tsv"), sep ="\t", quote = FALSE, row.names = FALSE, col.names = TRUE) 
```

## Differential expression analysis: identify difference among 5 CAR group of Tumor infiltrating T cells(TIL)
```{r}
Idents(all.CARs.combined) <- "Group"
levels(Idents(all.CARs.combined))

DefaultAssay(all.CARs.combined) <- "SCT"

options(future.globals.maxSize = 2000 * 1024^2)

BiCisCAR5_vs_other4CARs.de <- FindMarkers(all.CARs.combined, ident.1 = "CD28HTM.CD28z_CD8HTM.BBz_BiCisCAR5", ident.2 = c("FGFR4.CD28HTM.BBz_CAR1", "FGFR4.CD28HTM.CD28z_CAR2", "CD276.CD8HTM.BBz_CAR3",  "CD28HTM.BBz_CD8HTM.BBz_BiCisCAR4"), logfc.threshold = 0, verbose = FALSE)

 write.table(BiCisCAR5_vs_other4CARs.de, file = "../Data/allCARs.combined_SCT_BiCisCAR5_vs_other4CARs.DEGs.tsv", sep ="\t", quote = FALSE, row.names = TRUE, col.names = TRUE)
 
head(BiCisCAR5_vs_other4CARs.de, n = 20)



```

## EnhancedVolcano: publication-ready volcano plots with enhanced colouring and labeling
```{r}
options(download.file.method = "wininet")
BiocManager::install('EnhancedVolcano')
library(EnhancedVolcano)
```

# to show significantly differential genes in both sides
```{r}
top25_BiCisCAR5_vs_other4CARs.de <- top_n(BiCisCAR5_vs_other4CARs.de, n=25, wt =abs(avg_log2FC))
BiCisCAR5_vs_other4CARs_thres_ordered <- BiCisCAR5_vs_other4CARs.de %>% mutate(threshold = p_val_adj < 0.05) %>% arrange(p_val_adj)
BiCisCAR5_vs_other4CARs_thres_ordered$genelabels <- ""
BiCisCAR5_vs_other4CARs_thres_ordered[rownames(top25_BiCisCAR5_vs_other4CARs.de),]$genelabels <- "TRUE"

```

```{r}
genes.to.label = rownames(top25_BiCisCAR5_vs_other4CARs.de)

pdf("../SeuratAnalysisData/Data/DEGs_Plots/allCARs.combined_SCT_BiCisCAR5vsOther4CARs_DEGs_EnhancedVolcano.pdf", width = 10, height = 7)
EnhancedVolcano(BiCisCAR5_vs_other4CARs_thres_ordered,
                lab = rownames(BiCisCAR5_vs_other4CARs_thres_ordered),
                x = 'avg_log2FC',
                y = 'p_val_adj',
                selectLab = genes.to.label,
                xlab = bquote(~Log[2]~ 'fold change'),
                title = 'BiCisCAR5 versus other 4 CARs DEGs plot ',
                pCutoff = 10e-20,
                FCcutoff = 0.5,
                cutoffLineType = 'twodash',
                cutoffLineWidth = 0.8,
                 xlim = c(-1.1,1.1),
                pointSize = 2,
                labSize = 4.0,
                colAlpha = 1,
                legendLabels=c('Not sig.','Avg_Log2FC','p-value',
                               'p-value & Avg_Log2FC'),
                legendPosition = 'right',
                legendLabSize = 12,
                legendIconSize = 4.0,
                drawConnectors = TRUE,
                widthConnectors = 0.75)
dev.off()
```

##Heatmap show average top20 genes of DEGs btween  BiCisCAR5 and other 4 CARs
#Average RNA expression levels for each CAR replicate
```{r}
Idents(all.CARs.combined) <- "Group"
CARgroups_sct.averages <- AverageExpression(all.CARs.combined, assays = "SCT", slot = "data", return.seurat = TRUE, add.ident = "Group.IDs")

 saveRDS(CARgroups_sct.averages, file = "allCARs.combined_sct_AverageExpression_CARgroups.rds")
```

# sort the top20 differential expression genes between CAR5 and other 4 CAR; then Heatmap Plot
```{r}
top20_BiCisCAR5_vs_other4CARs.DEGs <- top_n(BiCisCAR5vsOther4CARs.DEGs, n=20, wt = avg_log2FC)

CARgroups_cols <- c("FGFR4.CD28HTM.BBz_CAR1" = "#FF6699", "FGFR4.CD28HTM.CD28z_CAR2"="#00CCCC", "CD276.CD8HTM.BBz_CAR3"="#9966FF",  "CD28HTM.BBz_CD8HTM.BBz_BiCisCAR4"="#CC6633", "CD28HTM.CD28z_CD8HTM.BBz_BiCisCAR5"="#0066FF")

DoHeatmap(CARgroups_sct.averages, features = top20_BiCisCAR5_vs_other4CARs.DEGs$Gene, size = 3, group.bar.height = 0.02, group.colors = CARgroups_cols,  draw.lines = FALSE) + scale_fill_gradientn(colors = rev(RColorBrewer::brewer.pal(n = 9, name = "RdBu")), na.value = "white") + guides(color=FALSE) + theme(axis.text.x = element_text(vjust = 0.25, angle = 90))

```

##change heatmap cols
```{r}
DoHeatmap(CARgroups_sct.averages, features = top20_BiCisCAR5_vs_other4CARs.DEGs$Gene, size = 3, group.bar.height = 0.02, group.colors = CARgroups_cols,  draw.lines = FALSE) + scale_fill_gradientn(colors = rev(RColorBrewer::brewer.pal(n = 9, name = "PiYG")), na.value = "white") + guides(color=FALSE) + theme(axis.text.x = element_text(vjust = 0.25, angle = 90))
```

##plot to show the differential expression between CARpos vs CARneg:
##Average expression for SCT data, split by Group.IDs_CARidents
```{r}
levels(all.CARs.combined$Group.IDs_CARidents)
Idents(all.CARs.combined) <- "Group"
Groups.IDs.CARidents_sct.averages <- AverageExpression(all.CARs.combined, assays = "SCT", slot = "data", return.seurat = TRUE, add.ident = "Group.IDs_CARidents")

 saveRDS(Groups.IDs.CARidents_sct.averages, file = "allCARs.combined_sct_AverageExpression_Groups.IDs.CARidents.rds")
```

#Heatmap Plot top20 DEGs
```{r}

Idents(Groups.IDs.CARidents_sct.averages) <- factor(all.CARs.combined2$Group.IDs_CARidents, levels = level_order4)

DoHeatmap(Groups.IDs.CARidents_sct.averages, features = top20_BiCisCAR5_vs_other4CARs.DEGs$Gene, size = 3, group.bar.height = 0.02, group.colors = CARgroups_cols,  draw.lines = FALSE) + scale_fill_gradientn(colors = rev(RColorBrewer::brewer.pal(n = 9, name = "RdBu")), na.value = "white") + guides(color=FALSE) + theme(axis.text.x = element_text(vjust = 0.25, angle = 90))
```

##Using SCT assay to Calculate cytotoxic score or proliferation score for each CAR
```{r}
cytotoxic.genes <- list(c("PRF1", "NKG7", "CCL3","SNHG25", "CCL5", "ATP5F1D", "COTL1", "CD8A", "GNLY", "PLEK", "CCL4", "GZMB", "GZMA", "SRSF9", "GZMH", "GZMK", "IFNG", "KLRD1", "RAC1", "HOPX"))

proliferation.genes <- list(c("TUBA1B", "TUBB", "RRM2", "TOP2A", "STMN1", "MKI67", "HMGB2", "DUT"))

all.CARs.combined <- AddModuleScore(all.CARs.combined, assay = "SCT", features = cytotoxic.genes, name = "cytotoxic.score")

all.CARs.combined <- AddModuleScore(all.CARs.combined, assay = "SCT", features = proliferation.genes, name = "proliferation.score")

head(all.CARs.combined[[]])
```

#### effector or cytokine genes
```{r}
cytokine.genes <- list(c("IL2", "IL4", "IL5", "IL13", "CXCL8", "IL9", "IL21", "EBI3", "IL21R","IL23R", 
                     "CCL3", "CCL4", "CCL5", "LTA", "IFNG", "TNF","GNLY", "GZMA", "GZMB", "GZMM",
                     "GZMK", "PRF1", "NKG7", "CSF2", "CTSW"))

effector.genes <- list(c("CCL3", "IFNG", "CCL4", "XCL1", "XCL2", "CSF2", "IL10", "HOPX", "LAG3", "PRF1",
                     "NKG7", "IL26"))
all.CARs.combined <- AddModuleScore(all.CARs.combined, features = cytokine.genes, name = "cytokine.score")

all.CARs.combined <- AddModuleScore(all.CARs.combined, features = effector.genes, name = "effector.score")
```

# plot cytotoxic or proliferation score for each CAR group, compare each CAR group
```{r}
library(ggpubr)

group_level_order <- c("FGFR4.CD28HTM.BBz_CAR1", "FGFR4.CD28HTM.CD28z_CAR2", "CD276.CD8HTM.BBz_CAR3",  "CD28HTM.BBz_CD8HTM.BBz_BiCisCAR4", "CD28HTM.CD28z_CD8HTM.BBz_BiCisCAR5")

CAR_comparisons <- list( c("FGFR4.CD28HTM.BBz_CAR1", "CD28HTM.CD28z_CD8HTM.BBz_BiCisCAR5"), c("FGFR4.CD28HTM.CD28z_CAR2", "CD28HTM.CD28z_CD8HTM.BBz_BiCisCAR5"), c("CD276.CD8HTM.BBz_CAR3", "CD28HTM.CD28z_CD8HTM.BBz_BiCisCAR5"), c("CD28HTM.BBz_CD8HTM.BBz_BiCisCAR4", "CD28HTM.CD28z_CD8HTM.BBz_BiCisCAR5") )

CARgroups_cols <- c("FGFR4.CD28HTM.BBz_CAR1" = "#FF6699", "FGFR4.CD28HTM.CD28z_CAR2"="#00CCCC", "CD276.CD8HTM.BBz_CAR3"="#9966FF",  "CD28HTM.BBz_CD8HTM.BBz_BiCisCAR4"="#CC6633", "CD28HTM.CD28z_CD8HTM.BBz_BiCisCAR5"="#0066FF")

as_tibble(all.CARs.combined[[]]) %>% 
  ggboxplot(x= "Group", y = "cytotoxic.score1",
          color = "Group", palette = CARgroups_cols,
          add = "jitter", add.params = list(size = 0.25))+
  rotate_x_text(angle = 45) +
  stat_compare_means(comparisons = CAR_comparisons)+
  stat_compare_means(label = "p.format")+ NoLegend()

as_tibble(all.CARs.combined[[]]) %>% 
  ggboxplot(x= "Group", y = "proliferation.score1",
          color = "Group", palette = CARgroups_cols,
          add = "jitter", add.params = list(size = 0.25))+
  rotate_x_text(angle = 45) +
  stat_compare_means(comparisons = CAR_comparisons)+
  stat_compare_means(label = "p.format")+ NoLegend()

```

##Violin plot with box plots inside
```{r}
as_tibble(all.CARs.combined[[]]) %>% 
  ggviolin(x = "Group", y = "cytotoxic.score1", fill = "Group",
         palette = CARgroups_cols,
         add = "boxplot", add.params = list(fill = "white"))+
  stat_compare_means(comparisons = CAR_comparisons, label = "p.format")+ 
  rotate_x_text(angle = 45)+ NoLegend()
   
```

##plot cytotoxic or cytokines score comparing CARpos vs CARneg
```{r}
as_tibble(all.CARs.combined[[]]) %>% 
  ggboxplot(x= "Group", y = "cytotoxic.score1",
          color = "CAR.idents", palette = "jco",
          add = "jitter",add.params = list(size = 0.25)) + 
  rotate_x_text(angle = 45) +
  stat_compare_means(aes(group = CAR.idents), label = "p.format")

as_tibble(all.CARs.combined[[]]) %>% 
  ggboxplot(x= "Group", y = "cytokine.score1",
          color = "CAR.idents", palette = "jco",
          add = "jitter",add.params = list(size = 0.25)) + 
  rotate_x_text(angle = 45) +
  stat_compare_means(aes(group = CAR.idents), label = "p.format")

as_tibble(all.CARs.combined[[]]) %>% 
  ggboxplot(x= "Group", y = "proliferation.score1",
          color = "CAR.idents", palette = "jco",
          add = "jitter",add.params = list(size = 0.25)) + 
  rotate_x_text(angle = 45) +
  stat_compare_means(aes(group = CAR.idents), label = "p.format")
```

##Vlnplot with boxPlot to plot multiple features
```{r}
top.DEGs_BiCisCAR5vsOthers <- c("GNLY", "GZMK", "GZMH", "TEX14", "NKG7", "FOS", "KLRD1", "CCL5")

plots <- VlnPlot(all.CARs.combined, top.DEGs_BiCisCAR5vsOthers, pt.size = 0, cols =CARgroups_cols, combine = FALSE)

for(i in 1:length(plots)) {
  plots[[i]] <- plots[[i]] + geom_boxplot(width=0.15, fill="white") + theme(legend.position = 'none')
}

CombinePlots(plots)

```






Add a new chunk by clicking the *Insert Chunk* button on the toolbar or by pressing *Ctrl+Alt+I*.

When you save the notebook, an HTML file containing the code and output will be saved alongside it (click the *Preview* button or press *Ctrl+Shift+K* to preview the HTML file).

The preview shows you a rendered HTML copy of the contents of the editor. Consequently, unlike *Knit*, *Preview* does not run any R code chunks. Instead, the output of the chunk when it was last run in the editor is displayed.
